src/Cli/dotnet/commands/dotnet-sln/add/Program.cs

```diff
@@ -2,143 +2,125 @@
-        {
-            var solutionFolders = new List<string>();
-
-            if (!IsPathInTreeRootedAtSolutionDirectory(projectFilePath))
-                return solutionFolders;
-
-            var currentDirString = $".{Path.DirectorySeparatorChar}";
-            if (projectFilePath.StartsWith(currentDirString))
+            catch (Exception ex)
             {
-                projectFilePath = projectFilePath.Substring(currentDirString.Length);
+                if (ex is SolutionException || ex.InnerException is SolutionException)
+                {
+                    throw new GracefulException(CommonLocalizableStrings.InvalidSolutionFormatString, solutionFileFullPath, ex.Message);
+                }
+                throw new GracefulException(ex.Message, ex);
             }
-
-            var projectDirectoryPath = TrimProject(projectFilePath);
-            if (string.IsNullOrEmpty(projectDirectoryPath))
-                return solutionFolders;
-
-            var solutionFoldersPath = TrimProjectDirectory(projectDirectoryPath);
-            if (string.IsNullOrEmpty(solutionFoldersPath))
-                return solutionFolders;
-
-            solutionFolders.AddRange(solutionFoldersPath.Split(Path.DirectorySeparatorChar));
-
-            return solutionFolders;
         }
 
-        private IList<string> DetermineSolutionFolder(SlnFile slnFile, string fullProjectPath)
+        private async Task AddProjectsToSolutionAsync(string solutionFileFullPath, string[] projectPaths, CancellationToken cancellationToken)
         {
-            if (_inRoot)
+            ISolutionSerializer serializer = SlnCommandParser.GetSolutionSerializer(solutionFileFullPath);
+            SolutionModel solution = await serializer.OpenAsync(solutionFileFullPath, cancellationToken);
+            // set UTF8 BOM encoding for .sln
+            if (serializer is ISolutionSerializer<SlnV12SerializerSettings> v12Serializer)
             {
-                // The user requested all projects go to the root folder
-                return null;
+                solution.SerializerExtension = v12Serializer.CreateModelExtension(new()
+                {
+                    Encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: true)
+                }); 
             }
-
-            if (_relativeRootSolutionFolders != null)
+            SolutionFolderModel solutionFolder = (!_inRoot && _solutionFolderPath != null)
+                ? solution.AddFolder(GetSolutionFolderPathWithForwardSlashes())
+                : null;
+            foreach (var projectPath in projectPaths)
             {
-                // The user has specified an explicit root
-                return _relativeRootSolutionFolders;
+                // Get full project path
+                var relativePath = Path.GetRelativePath(Path.GetDirectoryName(solutionFileFullPath), projectPath);
+                try
```

Comment:
I like that except that if we do that, it would cause the whole method to throw, which might be inconvenient if there are several projects that we want to add there